File: frontend/src/lib/api/auditTrail.ts
import { apiClient } from './client'
import { pageParams, readPageMeta, type PageMeta, type PageRequest } from './pagination'

/**
 * Trilha de auditoria — o cliente de `GET /api/v1/audit_trail` (`FE-446`).
 *
 * A trilha é o `paper_trail` (DEC-59): **uma** tabela `versions`, não uma tabela
 * por domínio. O legado tinha `GET /api/v1/trackings`; ela não é portada como
 * rota, e a evidência está no relatório da S19 — as colunas que aqueles filtros
 * usavam (`target_id`, `target_group_*`) nunca eram escritas por nenhum dos 20
 * emissores, e nenhuma tela do legado chamava o endpoint.
 *
 * **Quem lê (DEC-77):** a trilha global é de OG e Admin. O servidor devolve 403
 * para os demais; a tela não deve inferir isso do papel, deve tratar o 403.
 *
 * **Datas em ISO-8601** nos dois sentidos (`FE-440`). Nada de `dd/mm/aaaa` no
 * payload: a formatação acontece só na camada de apresentação.
 */

/** Uma pessoa citada pela trilha. `name` pode vir nulo se a conta foi apagada. */
export interface AuditActor {
  id: string
  name: string | null
  email: string | null
}

export interface AuditVersion {
  id: number
  item_type: string
  item_id: string
  /** Rótulo pt-BR do tipo, vindo do catálogo do servidor. */
  entity_label: string
  event: AuditEvent
  /**
   * A frase em pt-BR. Vem **pronta do servidor**, de propósito: é montada uma
   * vez só, no catálogo (`config/locales/pt-BR.yml`). Montá-la também aqui
   * seria a segunda implementação da mesma frase.
   */
  summary: string
  /** Autor REAL do ato. Em impersonação é quem impersonou, nunca o impersonado. */
  author: AuditActor | null
  /** Preenchido só em impersonação: quem a sessão aparentava ser. */
  impersonated: AuditActor | null
  reason: string | null
  ip_address: string | null
  /** ISO-8601 UTC. */
  occurred_at: string
  /** `campo: [antes, depois]`. */
  changes: Record<string, [unknown, unknown]>
  /** Foto completa do estado anterior — só no detalhe (DEC-78). */
  snapshot?: Record<string, unknown>
}

export type AuditEvent =
  | 'create'
  | 'update'
  | 'destroy'
  | 'impersonate_start'
  | 'impersonate_stop'

export interface AuditTrailFilters extends PageRequest {
  itemType?: string
  itemId?: string
  /** Id do autor real. */
  whodunnit?: string
  event?: AuditEvent
  /** ISO-8601. */
  from?: string
  /** ISO-8601. */
  to?: string
}

export interface AuditTrailPage {
  versions: AuditVersion[]
  meta: PageMeta
}

export interface AuditTypeOption {
  value: string
  label: string
}

function queryOf(f: AuditTrailFilters): Record<string, string | number> {
  const q: Record<string, string | number> = { ...pageParams(f) }
  if (f.itemType) q.item_type = f.itemType
  if (f.itemId) q.item_id = f.itemId
  if (f.whodunnit) q.whodunnit = f.whodunnit
  if (f.event) q.event = f.event
  if (f.from) q.from = f.from
  if (f.to) q.to = f.to
  return q
}

export const auditTrailApi = {
  /**
   * Índice da trilha. Os filtros são **combináveis**: cada um estreita o
   * anterior, e o total do envelope é o do filtrado.
   */
  list: async (filters: AuditTrailFilters = {}): Promise<AuditTrailPage> => {
    const resposta = await apiClient.getRaw<AuditVersion[]>('/api/v1/audit_trail', {
      params: queryOf(filters),
    })
    return {
      versions: resposta.data ?? [],
      meta: readPageMeta({ body: resposta.data, headers: resposta.headers as any }, filters),
    }
  },

  /** Detalhe, com a foto completa do estado anterior. */
  get: (id: number) => apiClient.get<AuditVersion>(`/api/v1/audit_trail/${id}`),

  /**
   * Vocabulário do filtro por tipo. Vem do servidor para a tela não precisar
   * conhecer a lista de models versionados — que é declarada num lugar só
   * (`Sfg::AuditTrail::VERSIONED`) e cresce a cada fatia.
   */
  types: () => apiClient.get<AuditTypeOption[]>('/api/v1/audit_trail/types'),
}

File: backend/app/controllers/api/v1/audit_trail.rb
# frozen_string_literal: true

module Api
  module V1
    # Leitura da trilha de auditoria — `BE-432` (listagem filtrável) e o detalhe.
    #
    # A trilha é o `paper_trail` (DEC-59): **uma** tabela `versions`, sem
    # `AuditEvent` e sem `trackings`. O payload é COMPLETO (DEC-78) — dá para
    # reconstruir o estado de qualquer registro em qualquer ponto do tempo.
    #
    # **`GET /api/v1/trackings` do legado não é portado como rota.** É esta rota.
    # O que o legado fazia aqui e não vem junto, com a evidência:
    #  - filtro por `target_id` e por `target_group_id`/`target_group_type`:
    #    **nenhum** dos 20 emissores do `TrackingFacade` escreve essas colunas;
    #  - `type`: coluna STI sempre `NULL`;
    #  - distância geográfica no detalhe (`BE-433`): descartada pelo **DEC-92**,
    #    que removeu a geolocalização inteira — a associação polimórfica não tem
    #    lado inverso nenhum no repositório de origem.
    #
    # **DEC-77 — quem lê.** A trilha GLOBAL (índice de tudo que aconteceu) é de
    # OG e Admin, e por isso é recurso próprio na matriz de autorização. O
    # histórico **do próprio objeto** é de quem vê o objeto, e por isso é
    # servido pelo endpoint do objeto, com `Sfg::AuditTrail.for_record`, e não
    # aqui.
    class AuditTrail < Grape::API
      helpers Api::V1::ControllerHelpers

      helpers do
        # Carrega os autores de uma página inteira num único SELECT. Sem isto a
        # entity faz um `find_by` por linha — 20 versões viram 40 consultas,
        # porque o impersonado também é usuário.
        def actors_for(versions)
          ids = versions.flat_map { |v| [v.whodunnit, v.impersonated_id] }.compact_blank.uniq
          return {} if ids.empty?

          ::User.where(id: ids).index_by { |u| u.id.to_s }
        end
      end

      namespace :audit_trail do
        before do
          authenticate_user!
          authorize!('audit_trail', :read)
        end

        desc 'Trilha de auditoria global' do
          summary 'Índice de versões'
          detail 'Paginação em cabeçalho. Filtros combináveis por tipo de registro, id, autor, evento e período.'
          success Api::Entities::AuditVersion
          failure [
            { code: 401, message: 'Não autenticado' },
            { code: 403, message: 'A trilha global é de OG e Admin (DEC-77)' }
          ]
        end
        params do
          optional :item_type, type: String, desc: 'Classe do registro (ex.: UserPermission)'
          optional :item_id, type: String, desc: 'Id do registro'
          optional :whodunnit, type: String, desc: 'Id do autor real do ato'
          # `impersonate_*` não vem do paper_trail: `Auth::ImpersonateService`
          # grava na MESMA tabela, porque a trilha é uma só (DEC-59). Sem
          # estarem aqui, o ato mais sensível do sistema não seria filtrável.
          optional :event, type: String,
                           values: %w[create update destroy impersonate_start impersonate_stop]
          # ISO-8601 na fronteira inteira (`FE-440`), e o tipo é `Sfg::Iso8601`
          # de propósito: o `DateTime` do Grape **aceita** `31/12/2025` e
          # `03/04/2026`, devolvendo 200 com a janela errada. Formato de data
          # ambíguo num filtro de auditoria é resposta errada com cara de certa.
          optional :from, type: DateTime, coerce_with: Sfg::Iso8601, desc: 'Início do período, ISO-8601'
          optional :to, type: DateTime, coerce_with: Sfg::Iso8601, desc: 'Fim do período, ISO-8601'
          optional :page, type: Integer, default: 1
          optional :per_page, type: Integer, default: 20
        end
        get '' do
          versions = paginate(Sfg::AuditTrail.filter(params))

          Api::Entities::AuditVersion.represent(
            versions.to_a, actors: actors_for(versions)
          )
        end

        # Tipos com verbete no catálogo, para a tela montar o filtro sem
        # conhecer a lista de models do servidor.
        desc 'Tipos de registro presentes na trilha' do
          summary 'Vocabulário do filtro por tipo'
        end
        get 'types' do
          Sfg::AuditSummary.known_types.map do |tipo|
            { value: tipo, label: Sfg::AuditSummary.label_for(tipo) }
          end
        end

        desc 'Detalhe de uma versão' do
          summary 'Uma versão, com a foto completa do estado anterior'
          detail 'O `snapshot` só aparece aqui: na listagem, 20 fotos completas seriam um download.'
          success Api::Entities::AuditVersion
          failure [
            { code: 401, message: 'Não autenticado' },
            { code: 403, message: 'A trilha global é de OG e Admin (DEC-77)' },
            { code: 404, message: 'Versão inexistente' }
          ]
        end
        params do
          requires :id, type: Integer, desc: 'Id da versão'
        end
        get ':id' do
          version = PaperTrail::Version.find_by(id: params[:id])
          error!({ error: 'not_found', message: I18n.t('platform.not_found') }, 404) if version.nil?

          Api::Entities::AuditVersion.represent(
            version, detail: true, actors: actors_for([version])
          )
        end
      end
    end
  end
end
